### ISSUE TITLE
Feature request: better handling of concurrent execution to prevent unnecessary out of memory failures 

### ISSUE THREAD

**Issue by alquraishi** [tensorflow/tensorflow#4552]
I have a model in which one operation happens to be very memory intensive and cannot be executed at once. I thought that by splitting the operation manually and executing parts of it individually (all on the GPU), I would get around the OOM problem. Unfortunately, it appears that TF aggressively tries to schedule as many ops to run in parallel as possible, even if that results in an OOM failure. I describe my specific issue further in this [Stack Overflow post](http://stackoverflow.com/questions/39643250/best-way-to-split-computation-too-large-to-fit-into-memory).

It seems to me that this problem should be reasonably easy to solve, and is rather a serious current limitation of TF. The basic principle is that a computation should not fail if it's possible, in some sequential order, to execute it in memory. I'm not talking about situations in which TF has to be clever about splitting a seemingly atomic operations in parts. That I understand can be quite difficult and requires "art" on the part of the programmer. What I'm describing is a situation in which the computation graph, _as described_, can in fact be executed without running out of memory, but because TF is (overly) aggressive in scheduling independent ops concurrently, it runs out of memory.

The solution seems rather straight-forward. Before putting yet another op onto the GPU (or whatever hardware), TF should check if that would cause the memory to run out. If so, it should check if there's a cyclic dependency on the current op getting executed, i.e. it would be impossible for it to move forward unless the current op is executed. If so then that's a legitimate problem and it would fail with an OOM error. On the other hand, if it's possible for it to wait until other ops are executed before trying again, then it should simply wait until more memory is available. I.e. it should have a way to run independent operations sequentially.

I'm not familiar enough with the internals of TF to know how difficult a change like this would be, but as it currently stands, it seems like it's preventing a large swath of models from getting executed that otherwise could get executed.

On a related note, the dynamic_rnn op already seems to do this, where it's able to shuffle memory back and forth until the computation is done, and so perhaps there's an existing partial solution that I'm not aware of?

**Comment by alquraishi** [comment]
Quick update. I discovered that using `map_fn` and setting `parallel_iterations` to a small number addresses my current needs. Nonetheless, I think it would be very helpful if there was a general mechanism by which TF prevented (unnecessary) failures due to running out of memory.

**Comment by yaroslavvb** [comment]
@yuanbyu was working on related stuff. The issue is that independent ops schedule to execute concurrently even if that exceeds device memory budget

**Comment by jmchen-g** [comment]
@yuanbyu Could you take a look at this? Thanks.

**Comment by aselle** [comment]
#2126 is a catch all for improvements to TensorFlow placement. I'm closing this issue and redirecting this query there. Thanks.

**Comment by yaroslavvb** [comment]
Memory efficient execution is an open research area, and it's good to have a concrete application in mind to have a practical solution. For instance I found that "out of memory" was often caused by bad execution order, rather than executing too many of them concurrently.

IE consider computational graph below
![caterpillar](https://cloud.githubusercontent.com/assets/23068/22432653/7dc7bf20-e6ca-11e6-879c-14880a4d26d3.png)

TF schedules things as soon as they are ready to execute, so if "square" op takes long to compute, it'll compute all the "circles", and then keep them in memory, and then compute the squares. This needs O(n) memory where `n` is the length of chain, whereas memory-optimal execution need O(1) memory and will go "circle->square->circle->square".

The general problem is known in computer science as "one-shot black pebbling" first formalized in http://graal.ens-lyon.fr/~lmarchal/scheduling/sethi_complete_register_allocation.pdf

It is provably [hard](https://arxiv.org/abs/1109.4910) to compute, but things can look better for concrete models.

Another set of memory-constrained applications can be enabled by discarding some intermediate values and recomputing them later. That's corresponds to dropping "one-shot" in the "one-shot pebbling game", and in deep learning this comes up in approaches like training neural networks where gradients don't fit in [memory](https://arxiv.org/abs/1604.06174)

Regarding your proposal, it sounds like you are proposing to add memory swapping to more ops in TensorFlow?

This is what the implementation does with it right now:
https://github.com/tensorflow/tensorflow/blob/64edd34ce69b4a8033af5d217cb8894105297d8a/tensorflow/core/kernels/stack_ops.cc#L215

There's a trade off between transfer and recomputation. For instance it may be 7-10x faster to recompute a cheap op like `mul` or `concat` on GPU rather than fetch previous result from main memory.

To summarize, you can reduce memory by picking a better execution order, by recomputing parts of the graph on demand, or by swapping things out to GPU memory. Can you be more specific with the model you are trying to run to make sure that swapping is indeed the correct thing here?

**Comment by alquraishi** [comment]
I think we may be talking about two distinct levels of abstractions. I understand that "solving" the time-optimal or space-optimal memory execution order problem may be difficult. What I'm asking for is something that would prevent TF from executing an op if that would result in an OOM error. Perhaps I'm making an undue assumption: I'm assuming that TF can estimate the memory requirement of a single op. If that's the case, then, given the current state of available memory, it should be able to determine whether executing another op will lead to an OOM or not. If it would, then my desired behavior would be that it just waits. I.e. it sets up a schedule that enforces sequentiality if need be (with some timeout heuristics, etc) to prevent OOM errors.

Or are you saying that it already does that, but finds itself in an unrecoverable state where computing the next necessary op is simply impossible? I'm somewhat doubtful that's the case, at least in my current graph.

**Comment by vrv** [comment]
@alquraishi it wasn't easy to determine whether executing an op would OOM in the current framework, since you're already in the execution of the OpKernel when you find out you run out of memory, and the framework doesn't really handle re-executing an OpKernel well (it may be possible, though).

I think there are two possible approaches to solve this (none of which I'm too involved in at the moment, but something I've thought about):

* Once we get C++ shape inference in the runtime (right now it just backs the python shape inference framework), you would be able to know at runtime the approximate output memory requirements of every operation, and so you could indeed know ahead of time whether an operation could run.  This wouldn't include temporary memories required by libraries like cudnn), but it would probably be the low hanging fruit optimization.

* XLA's compiled approach needs to know the sizes of tensors at compile time, providing the information to the memory and instruction scheduler for XLA to possibly do the right thing.  I'm more optimistic about this approach working well, since the standard executor is very dynamic, and XLA might have a better time of accomplishing this goal. @tatatodd  might have more to say on this.

* It would be also nice to have the recompuatation/memory use tradeoff in the above solutions.

**Comment by yaroslavvb** [comment]
@alquraishi estimating memory needed is hard, but simple heuristic might be to use an upper bound instead of true memory requirement. IE, suppose the largest op output is 1GB. Then you assume that once memory hits 11GB on TitanX, any new op executed will cause OOM and output needs to be placed on CPU.

Hence, once you hit 11GB, all GPU kernels place their output in main RAM until GPU RAM is reclaimed. This will avoid OOM crash at the expense of introducing a bunch of GPU->CPU and CPU->GPU memory copies needed to run GPU kernels.

This is a "run-time" approach mirroring what dynamic_rnn does. A competing approach is static graph analysis which would determine which ops are likely to hit this scenario, and pin those ops to CPU device. This means CPU kernel implementation will be used, and extra GPU<->CPU memory copies are avoided.

The later could be done purely on client side, using `tf.device` annotations

**Comment by alquraishi** [comment]
Yes these would all be reasonable solutions. The problem I'm having right now is that I'm unable to run the model _at all_, which is a big step down from having to do some extra GPU<->CPU copying. I've tried to manually enforce sequential computations using `control_dependencies`, and it's alleviated the problem somewhat, but there are still many models I can't run that I would be able otherwise.

**Comment by girving** [comment]
Closing since I think the issue as-is is too open ended.  Would be great to improve the situation, but I don't think leaving the issue open will encourage the necessary research.
